token-lending/program/src/instruction.rs

```diff
@@ -217,12 +217,35 @@ pub enum LendingInstruction {
     ///   9. `[]` Token program id.
     ///   10 `[optional, writable]` Host fee receiver account.
     BorrowObligationLiquidity {
-        /// Amount of liquidity to borrow - u64::MAX for 100% of borrowing power
+        /// Amount of liquidity to borrow as a whole number
         liquidity_amount: u64,
-        // @TODO: slippage constraint - https://git.io/JmV67
     },
 
     // 11
+    /// Borrow as much liquidity as possible from a reserve by depositing collateral tokens.
+    /// Requires a refreshed obligation and reserve.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` Source borrow reserve liquidity supply SPL Token account.
+    ///   1. `[writable]` Destination liquidity token account.
+    ///                     Minted by borrow reserve liquidity mint.
+    ///   2. `[writable]` Borrow reserve account - refreshed.
+    ///   3. `[writable]` Borrow reserve liquidity fee receiver account.
+    ///                     Must be the fee account specified at InitReserve.
+    ///   4. `[writable]` Obligation account - refreshed.
+    ///   5. `[]` Lending market account.
+    ///   6. `[]` Derived lending market authority.
+    ///   7. `[signer]` Obligation owner.
+    ///   8. `[]` Clock sysvar.
+    ///   9. `[]` Token program id.
+    ///   10 `[optional, writable]` Host fee receiver account.
+    BorrowAllObligationLiquidity {
+        /// Minimum amount of liquidity to receive as a whole number
+        slippage_limit: u64,
+    },
+
+    // 12
```

Comment:
This changes the index of all the subsequent instructions, a breaking change. I think new instructions should be appended?

However, I _think_ I disagree that this should be a separate instruction at all, since `liquidity_amount == u64::MAX` is already special-cased in the implementation.

This makes it possible to call `BorrowObligationLiquidity` with `u64::MAX` and it has different behavior than `BorrowAllObligationLiquidity`  while having the same semantic intent.

Reply:
> I didn't think we needed to worry about breaking changes for a reference implementation

This is reasonable, I don't have a strong opinion, and it does keep the code cleaner this way. The only concern I have is that it makes it harder for downstream forks, because it's a breaking change for them. But I don't know if any of the major ones that started as forks (Solend, Port, Larix, Tulip, etc) actually pull anything from here.

> The idea is that special-casing `u64::MAX` at the borrow instruction is unclear.

Also reasonable, this design choice is owing to me not knowing what I'm doing so trying to keep things simple 🙃

Reply:
> But I don't know if any of the major ones that started as forks (Solend, Port, Larix, Tulip, etc) actually pull anything from here.

Ah right, in that case, we can do them a solid and add it to the end.

> Also reasonable, this design choice is owing to me not knowing what I'm doing so trying to keep things simple

Too modest! What do you think is the best approach to make this clearer? Note that "keeping it as it is" is a valid answer :smile:

Reply:
> we can do them a solid and add it to the end.

Side note -- is there a way to manually set the enum "keys" so that it can appear lexically in order but use the index at the end?

> What do you think is the best approach to make this clearer?

I like _the idea_ of not using `u64::MAX` if there's a better option, though it's probably not worth changing. Keep it as is I guess!

I personally prefer the original implementation without adding an instruction:
https://github.com/jmalcic/solana-program-library/blob/63b6e9a307487c86c06f5b9a8fb5ecb846d89602/token-lending/program/src/processor.rs#L1153-L1156

Since this is the only business logic change, adding an instruction + creator to get to the same processor code path is a bigger change than just adding an arg to the same instruction.

I would probably approach this by using the same instruction, and making this optional: https://github.com/jmalcic/solana-program-library/blob/63b6e9a307487c86c06f5b9a8fb5ecb846d89602/token-lending/program/src/instruction.rs#L375

Parse a u64 if it's there, and if not, default to 0. This makes it a nonbreaking change for the instruction data.

Reply:
Okay, I've reverted the splitting of the instruction and made the parsing of the `u64` optional. https://github.com/solana-labs/solana-program-library/blob/ede3bdfdc993b07ec7fc5fb126f41253b117b680/token-lending/program/src/instruction.rs#L375